sympy/interactive/session.py

```diff
@@ -85,6 +90,76 @@ def _make_message(ipython=True, quiet=False, source=None):
     return message
 
 
+class IntegerWrapper(ast.NodeTransformer):
+    """Wraps all integers in a call to Integer."""
+
+    def visit_Num(self, node):
+        if isinstance(node.n, int):
+            return ast.Call(func=ast.Name(id='Integer', ctx=ast.Load()),
+                            args=[node], keywords=[],
+                            starargs=None, kwargs=None)
+        return node
+
+    def visit_Call(self, node):
+        if isinstance(node.func, ast.Name) and (node.func.id == "Integer"
+                or node.func.id == "int"):
+            return node
+        return self.generic_visit(node)
+
+
+class AutomaticSymbols(ast.NodeTransformer):
+    """Add missing Symbol definitions automatically."""
+
+    def __init__(self, shell):
+        super(AutomaticSymbols, self).__init__()
+        self.shell = shell
+        self.names = []
+
+    def visit_Module(self, node):
+        ignored_names = list(self.shell.user_ns.keys()) + dir(builtins)
+
+        for s in node.body:
+            self.visit(s)
+
+        for v in self.names:
+            if v in ignored_names:
+                continue
+
+            assign = ast.Assign(targets=[ast.Name(id=v, ctx=ast.Store())],
+                                value=ast.Call(func=ast.Name(id='Symbol',
+                                                             ctx=ast.Load()),
+                                               args=[ast.Str(s=v)], keywords=[],
+                                               starargs=None, kwargs=None))
+            node.body.insert(0, assign)
+
+        newnode = ast.Module(body=node.body)
+        ast.copy_location(newnode, node)
+        ast.fix_missing_locations(newnode)
+        return newnode
+
+    def visit_Name(self, node):
+        if isinstance(node.ctx, ast.Load):
+            self.names.append(node.id)
+        return node
+
+
+class FloatRationalizer(ast.NodeTransformer):
+    """Wraps all floats in a call to Rational."""
+
+    def visit_Num(self, node):
+        if isinstance(node.n, float):
+            return ast.Call(func=ast.Name(id='Rational', ctx=ast.Load()),
+                            args=[ast.Str(s=repr(node.n))], keywords=[],
```

Comment:
Is it possible to use the exact string form? If you enter a float that has more precision than a Python float, like `1.00000000000000000001`, it gets rounded before being wrapped.

Reply:
Not with ast.  It sees not a "string form", but a float literal.  Which is 1.0 in this case.

Reply:
I think you may be able to "find" the original float by tokenizing the input string and looking for the literal with the appropriate col_offset and lineno. I don't know if it will be robust, though.

Reply:
I'm not suggesting to tokenize - transform - untokenize (as we did before with Integer wrapping). That has issues, such as changing the code that's shown in tracebacks. You still use an ast transformer, but instead of using node.n, you use tokenize the input string, then match up the token that has the corresponding node.col_offset and lineno.  This is assuming the full original input string is available from the IPython API.

Reply:
For example:

```py
>>> ast.parse('2 + 1.0000000000000001').body[0].value.right.lineno
1
>>> ast.parse('2 + 1.0000000000000001').body[0].value.right.col_offset
4
>>> for token in tokenize.tokenize(io.BytesIO('2 + 1.0000000000000001'.encode('utf-8')).readline):
...     print(token)
TokenInfo(type=59 (BACKQUOTE), string='utf-8', start=(0, 0), end=(0, 0), line='')
TokenInfo(type=2 (NUMBER), string='2', start=(1, 0), end=(1, 1), line='2 + 1.0000000000000001')
TokenInfo(type=53 (OP), string='+', start=(1, 2), end=(1, 3), line='2 + 1.0000000000000001')
TokenInfo(type=2 (NUMBER), string='1.0000000000000001', start=(1, 4), end=(1, 22), line='2 + 1.0000000000000001')
TokenInfo(type=0 (ENDMARKER), string='', start=(2, 0), end=(2, 0), line='')
```

So (1, 4) corresponds to the token `TokenInfo(type=2 (NUMBER), string='1.000000000001', start=(1, 4), end=(1, 18), line='2 + 1.000000000001')`, which has the original string

Reply:
@asmeurer The input string is not passed to `run_ast_nodes` nor available to AST transformers. We'd have to monkey-patch `run_cell`, but I'm afraid there is a lot of processing between `run_cell` call and AST transformation.

Reply:
@asmeurer I'm not sure why showing transformed input in the traceback is a problem. IPython already does that for string-based transformers and line magics, and it may be more informative to users.

Reply:
> You still use an ast transformer, but instead of using node.n, you use tokenize the input string

This may work only if NodeTransformer will not shift original col_offset/lineno after modifications.

> We have to monkey-patch run_cell

No.  You just can use other available IPython transformations, string-based.  And parse & run ast transformation here, together with tokenization.

But IMHO, this way just overkill c.f. simple tokenize->fix->untokenize.

> I'm not suggesting to tokenize -> transform -> untokenize (as we did before with Integer wrapping). That has issues, such as changing the code that's shown in tracebacks.

I bet, it doesn't related to using tokenize, but to your monkey-patched run_cell() method.  You should use provided by IPython API, namely string-based input (registered in python_line_transforms).

Reply:
If a tokenize based transformer is no worse than an ast transformer then why are we changing to ast transformers in the first place?

Reply:
Good question for PR owner.  But lets make it clear: currently sympy doesn't use **any** IPython API for transformers.  Nor ast transformers, nor string-based.

Reply:
@asmeurer As you said, IPython ast transformers do not show transformed input in the traceback–though I’m not sure it’s neccessary. This PR aims to use IPython APIs for better compatibility, since as @skirpichev said, it currently monkey-patches IPython.

Reply:
Also, if we use string-based or token-based transformers, maybe we can hook existing transformers at `sympy.parsing.sympy_parser` like `convert_xor`.

Reply:
How does IPython handle its extensions of Python (like magics), which can appear inside of Python statements, like

```
for i in range(10):
    %time i + 1
```

Does IPython have a special Python parser, or is it simpler than that? 

I agree that an xor extension would also be useful, and it has to happen at tokenization, before the Python AST is created, due to they way ^ is parsed (it has a different precedence than **, so `x + y^2` would be `Xor(Add(x, y), 2)` in the AST).

Reply:
@asmeurer IPython just processes line by line, and look for magic commands starting with percent signs, at the start of line and preprocesses it before passing to Python.